### ISSUE TITLE
Redirection Issue? - Or just native Android behaviour?

### ISSUE THREAD

**Issue by paulwillen** [openid/AppAuth-Android#408]
Hi Sorry for clogging up the other issue (https://github.com/openid/AppAuth-Android/issues/245)

So to answer the questions you there asked:
what IDP / OAuth2 implementation are you integrating with? 
Answer: A Laravel Passport implementation

Does the redirect occur automatically, i.e. as a JS redirect, or in response to user interaction such as clicking on a button?

Situation: I'm authorized
Action: I click the authorize button again; wich makes the call to the authorization url. This one sees that I'm already authorized and redirects me back to the redirect-url. That process is arranged server-side I assume (the redirect); and thus I don't have control over that. I always get redirected to the web-page and never to the app again (because I guess it's not a manual action). 

If it's the case that we're looking into a security issue of android here: Any idea how to solve this?

**Comment by tikurahul** [comment]
I think in your case the best thing is to navigate to an intermediate `redirectUrl` (that you control) and something that the user can click to return to the app.

**Comment by iainmcgin** [comment]
Yes, this is the classic problem that many users of AppAuth encounter with automatic redirects that occur if the user has already consented to the scopes the app is requesting. 

Looking at Laravel's documentation, it does not appear that they support the `prompt=consent` request option, which some IDPs respect and forces re-consent, resulting in a redirect based on user input, which should work.

The alternative is as Rahul states: you will need an intermediary web page to capture the redirect; I have a demo of this [here](https://appauth.demo-app.io/oauth2redirect); if you look at the source, what it does is take the request parameters to the page and send those to a custom scheme after the user clicks on a "return to app" button:

```js
var redirectUri = "net.openid.appauthdemo://oauth2redirect";
    
// copy query and fragment verbatim to the redirect uri, if they
// are defined
if (window.location.search) {
    redirectUri += window.location.search;
}
if (window.location.hash) {
    redirectUri += window.location.hash;
}

document.getElementById('redirect-link').onclick = () => {
    window.location.href = redirectUri;
};
```

Using a technique like this will allow your app to capture redirects from the IDP directly if the user is presented with a consent flow, and in the event of an automatic redirect, gives you a way to recover via the interstitial page. This page could show a friendly "welcome back!" style message so that it does not seem out of place.
